cpp/src/join/conditional_join.cu

```diff
@@ -35,6 +35,103 @@
+  if (right_num_rows == 0) {
+    switch (join_type) {
+      case join_kind::LEFT_ANTI_JOIN:
+        return std::make_unique<rmm::device_uvector<size_type>>(left_num_rows, stream, mr);
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr);
+      default: CUDF_FAIL("Invalid join kind."); break;
+    }
+  } else if (left_num_rows == 0) {
+    switch (join_type) {
+      case join_kind::LEFT_ANTI_JOIN:
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr);
+      default: CUDF_FAIL("Invalid join kind."); break;
+    }
+  }
+
+  auto const has_nulls = binary_predicate.may_evaluate_null(left, right, stream);
+
+  auto const parser =
+    ast::detail::expression_parser{binary_predicate, left, right, has_nulls, stream, mr};
+  CUDF_EXPECTS(parser.output_type().id() == type_id::BOOL8,
+               "The expression must produce a boolean output.");
+
+  auto left_table  = table_device_view::create(left, stream);
+  auto right_table = table_device_view::create(right, stream);
+
+  detail::grid_1d const config(left_num_rows, DEFAULT_JOIN_BLOCK_SIZE);
+  auto const shmem_size_per_block = parser.shmem_per_thread * config.num_threads_per_block;
+
+  // the code below can also be taken out in the context of semi & anti, but
+  // i will leave that for another PR for sake of being conservative
+  std::size_t join_size;
+  if (output_size.has_value()) {
+    join_size = *output_size;
+  } else {
+    rmm::device_scalar<std::size_t> size(0, stream, mr);
+    if (has_nulls) {
+      compute_conditional_join_output_size<DEFAULT_JOIN_BLOCK_SIZE, true>
+        <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
+          *left_table, *right_table, join_type, parser.device_expression_data, false, size.data());
+    } else {
+      compute_conditional_join_output_size<DEFAULT_JOIN_BLOCK_SIZE, false>
+        <<<config.num_blocks, config.num_threads_per_block, shmem_size_per_block, stream.value()>>>(
+          *left_table, *right_table, join_type, parser.device_expression_data, false, size.data());
+    }
+    join_size = size.value(stream);
+  }
+
+  if (join_size == 0) { return std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr); }
+
+  rmm::device_scalar<size_type> write_index(0, stream);
+
+  auto left_indices  = std::make_unique<rmm::device_uvector<size_type>>(join_size, stream, mr);
+  auto right_indices = std::make_unique<rmm::device_uvector<size_type>>(join_size, stream, mr);
+
+  auto const& join_output_l = left_indices->data();
+  // i am allocating twice the default cache size for these joins, because, you aren't concerned
+  // with the right side at all. So that right side is more space to be used for the left side. This
+  // should increase the size of the batches, leading to less frequent updates to global memory
+  // which should also improve speed
```

Comment:
I am not super familiar with the cache size parameter here. Can you outline how this parameter is used, why you chose a factor of 2, and whether this choice of value is supported by benchmark data? Also I would recommend declaring a named value like `auto constexpr cache_size_factor = 2` and using that instead.

Also can you write this comment with complete sentences / proper grammar and capitalization?

Reply:
(I need to refresh my memory on this PR 😅 )Yeah so this is how the cache is used in the original kernel
```
  __shared__ cudf::size_type join_shared_l[num_warps][output_cache_size];
  __shared__ cudf::size_type join_shared_r[num_warps][output_cache_size];
  ```
  Since we remove the right side from shared memory, we will be allocating half the amount of memory we did in shared memory. That is more memory for the left side to use, which is why I multiplied it by 2. The amount of memory allocated to shared memory should still remain the same. 
  
I was hoping to reap some benefit alongside experiment to see if this would improve performance. IIRC, `flush_output_cache` batch write size to global memory would be same as the cache size used here. By doubling it, that write size would increase. Unfortunately from the benchmarking I did on my GPU, I did not notice anything significant... Gains in the realm of 1-3%. I need to take another look at the evaluator for intermediate expressions, I can't recall if it had some influence to all of this. I don't remember how the implementation worked exactly. 

I am thinking I'll need to run the benchmark through ncompute and seeing if there is any substance to justifying this change. Hopefully google benchmarks and ncompute will get along with eachother... 😅 

Please lmk if I have explained poorly somewhere, keep in mind I'm still rather new to CUDA & C++ so I may be using incorrect nomenclature somewhere.

Reply:
Twice the default size is unnecessary. It's a number chosen without many fine tunings IIRC.

Reply:
Sounds good, I'll revert this